src/passes/MultiMemoryLowering.cpp

```diff
@@ -0,0 +1,394 @@
+    combinedMemory = Names::getValidMemoryName(*wasm, "combined_memory");
+  }
+
+  void addOffsetGlobals() {
+    auto addGlobal = [&](Name name, size_t offset) {
+      auto global =
+        Builder::makeGlobal(name,
+                            pointerType,
+                            Builder(*wasm).makeConst(
+                              pointerType == Type::i32
+                                ? Literal::makeFromInt32(offset, pointerType)
+                                : Literal::makeFromInt64(offset, pointerType)),
+                            Builder::Mutable);
+      global->hasExplicitName = true;
+      wasm->addGlobal(std::move(global));
+    };
+
+    size_t offsetRunningTotal = 0;
+    for (Index i = 0; i < wasm->memories.size(); i++) {
+      auto& memory = wasm->memories[i];
+      memoryIdxMap[memory->name] = i;
+      // We don't need a page offset global for the first memory as it's always
+      // 0
+      if (i != 0) {
+        Name name = Names::getValidGlobalName(
+          *wasm, memory->name.toString() + "_page_offset");
+        offsetGlobalNames.push_back(std::move(name));
+        addGlobal(name, offsetRunningTotal);
+      }
+      offsetRunningTotal += memory->initial;
+    }
+  }
+
+  void adjustActiveDataSegmentOffsets() {
+    Builder builder(*wasm);
+    ModuleUtils::iterActiveDataSegments(*wasm, [&](DataSegment* dataSegment) {
+      assert(dataSegment->offset->is<Const>() &&
+             "TODO: handle non-const segment offsets");
+      auto idx = memoryIdxMap.at(dataSegment->memory);
+      dataSegment->memory = combinedMemory;
+      // No need to update the offset of data segments for the first memory
+      if (idx != 0) {
+        // We need to subtract 1 here because there is no offsetGlobal for the
+        // first memory
+        auto offsetGlobalName = getOffsetGlobal(idx);
+        dataSegment->offset = builder.makeBinary(
+          Abstract::getBinary(pointerType, Abstract::Add),
+          builder.makeGlobalGet(offsetGlobalName, pointerType),
+          dataSegment->offset);
+      }
+    });
+  }
+
+  void createMemorySizeFunctions() {
+    for (Index i = 0; i < wasm->memories.size(); i++) {
+      auto function = memorySize(i);
+      memorySizeNames.push_back(function->name);
+      wasm->addFunction(std::move(function));
+    }
+  }
+
+  void createMemoryGrowFunctions() {
+    for (Index i = 0; i < wasm->memories.size(); i++) {
+      auto function = memoryGrow(i);
+      memoryGrowNames.push_back(function->name);
+      wasm->addFunction(std::move(function));
+    }
+  }
+
+  // This function replaces memory.grow instruction calls in the wasm module.
+  // Because the multiple discrete memories are lowered into a single memory,
+  // we need to adjust offsets as a particular memory receives an
+  // instruction to grow.
+  std::unique_ptr<Function> memoryGrow(Index memIdx) {
+    Builder builder(*wasm);
+    Name name = "custom_memory_grow_" + std::to_string(memIdx);
```

Comment:
I don't think the "custom_" prefix adds much here. Maybe we can put the original memory name into the name instead like we do with the offset globals?
```suggestion
    Name name = "memory_grow_" + std::to_string(memIdx);
```

Reply:
Played around with a few combinations for name, I think memoryName + "_grow" is the most readable. dropping the memidx. Doing the same naming scheme for the replacement memory size function names.